napari/layers/base/base.py

```diff
@@ -329,6 +329,12 @@ def __init__(
             ),
         )
         self.name = name
+        self.events.connect(self._clean_cache)
```

Comment:
Generally I'd been trying to avoid us having these "internal" event connections (which is why the tests above were asserting no connections) and I don't think we need to use our event system here. Right now whenever we change colormaps or other unrelated things the cache will change.

The most time consuming operation is computing `_extent_data`, and that only changes when the data changes. So instead we can just store `_extent_data`, say as an actual value, not a property, and then added a `_get_extent_data` method that we just call to update that value when the data changes then we should be fine. That would result in the smallest number of calls. What do @andy-sweet @tlambert03 @jni think

Reply:
Also computing `Affine.scale` is time-consuming (1/4 of `Image._get_extent_world` time calculation).

I also found that in calculation of `extent` Affine object is constructed two times (two calls of `_data_to_world`).

![Zrzut ekranu z 2021-11-11 10-44-57](https://user-images.githubusercontent.com/3826210/141276021-5c50295b-c389-4536-8927-96560a0012f3.png)

Because of this tat is easy to introduce time consuming step in code, without notice from main maintainers I select option to add cache and refresh it on any change in data model. For me best option is to only use blacklist of signals which should not refresh cache. 

The core of problem is that still some "heavy" computations are hidden under properties. like `_data_to_world` which construct `Affine` object each time.

Reply:
> The core of problem is that still some "heavy" computations are hidden under properties. like _data_to_world which construct Affine object each time.

Ok, that's fair - I still think we should then just trigger setting the `_extent_data` inside the `scale` setter then rather than use the event system. I can try taking a look at this now

Reply:
I have created #3606 as an alternative

Reply:
I know it doesn't help with the general problems described here, but we could avoid computing `_data_to_world` twice fairly easily. For example, something like the following

```python
    @property
    def extent(self) -> Extent:
        """Extent of layer in data and world coordinates."""
        data = self._extent_data
        data_to_world = self._data_to_world
        return Extent(
            data=data,
            world=self._get_extent_world(data, data_to_world),
            step=abs(data_to_world.scale),
        )
```

where `_get_extent_world` is passed the computed `_data_to_world` (and so could be a static method or just a plain function).

Reply:
> I know it doesn't help with the general problems described here, but we could avoid computing _data_to_world twice fairly easily. For example, something like the following

yeah, we should do that too